### PR TITLE
New Unit Test to Detect DTO Model Changes

### PR DESCRIPTION
**Pull request by mwestacott**
### JIRA link

None

### Change description

* Now, but especially after 'Save & Resume' for registrations has gone LIVE, we need to be very careful with regards changing the API OE submission data model.
* This new test should fail if a DTO structure incompatibility is detected.
* Instructions of what to do added to the Javadoc of the testCanGenerateDtoModelFromValidJson() method.

### Work checklist

- [X] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+     * @throws Exception If DTO model structure incompatibility detected
+     */
+    @Test
+    public void testCanGenerateDtoModelFromValidJson() throws Exception {
+
+        TestUtils testUtils = new TestUtils();
+        File file = testUtils.getFile("overseas_entity_valid.json");
+        String jsonData = testUtils.readFileReturnString(file);
+
+        getObjectMapper().readValue(jsonData, OverseasEntitySubmissionDto.class);
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java]: wondering if you could use an assertDoesNotThrow here - https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#assertDoesNotThrow(org.junit.jupiter.api.function.Executable)

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    @Test
+    public void testCannotGenerateDtoModelFromInvalidJson() throws Exception {
+        TestUtils testUtils = new TestUtils();
+        File file = testUtils.getFile("overseas_entity_invalid.json");
+        String jsonData = testUtils.readFileReturnString(file);
+
+        try {
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:47]: maybe try an assertThrows here, i guess you should be able to have an error message on it too

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        File file = testUtils.getFile("overseas_entity_valid.json");
+        String jsonData = testUtils.readFileReturnString(file);
+
+        OverseasEntitySubmissionDto dto = getObjectMapper().readValue(jsonData, OverseasEntitySubmissionDto.class);
+
+        assertNotNull(dto);
+    }
+
+    @Test
+    void testCannotGenerateDtoModelFromInvalidJson() throws Exception {
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:42]: this test might not be needed as we really only want to test that old json can be converted into latest Dto

Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:42]: Agree. Will remove.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/TestUtils.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        try {
+            buff = new BufferedReader(
+                    new FileReader(filename));
+            String line;
+            while ((line = buff.readLine()) != null) {
+                builder.append(line);
+            }
+        } catch (Exception e) {
+            LOGGER.severe("Parsing json file error: " + e.getMessage());
+        }
```
[the remark this answers is not in the document]
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/TestUtils.java:27]: Copied this from chips-filing-consumer code, but yes, I do. Even better, will change to use a try-with-resources statement.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java (hunk 1 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java]: Went for an assertNotNull() - satisfies Sonar and exception couldn't have been thrown if it reaches that line.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java (hunk 2 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/model/DtoModelChangeTest.java:47]: Method deleted so no action here.

**Review by markpit — APPROVED** [review]

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/TestUtils.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.logging.Logger;
+
+/**
+ *  Utilities to read test resources.
+ */
+public final class TestUtils {
+
+    private static final Logger LOGGER = Logger.getLogger(TestUtils.class.getName());
+
+    public String readFileReturnString(File filename) {
```
Comment (AndreKramer7654) [inline src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/TestUtils.java:15]: rename to readString(filename)? Java 11 has File.readString()

**Review by AndreKramer7654 — APPROVED** [review]
one optional name change suggestion but otherwise valid test.

File: src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/TestUtils.java (hunk 5 above)
Comment (mwestacott) [reply src/test/java/uk/gov/companieshouse/overseasentitiesapi/utils/TestUtils.java:15]: Yes, agree (copied this code from somewhere else so should have changed that method name).

**Pull request merged**